sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -839,8 +839,12 @@ object DateTimeUtils {
    * is valid for the `toZone` time zone, thus the local date-time may be adjusted.
    */
   def convertTz(micros: Long, fromZone: ZoneId, toZone: ZoneId): Long = {
-    val rebasedDateTime = getLocalDateTime(micros, toZone).atZone(fromZone)
-    instantToMicros(rebasedDateTime.toInstant)
+    val localDateTime = toJavaTimestamp(micros).toLocalDateTime
+    val zonedDateTimeFromZone = ZonedDateTime.of(localDateTime, fromZone)
+    val zonedDateTimeToZone = zonedDateTimeFromZone.withZoneSameInstant(toZone)
+    val localDateTimeConverted = zonedDateTimeToZone.toLocalDateTime
+    val timestamp = Timestamp.valueOf(localDateTimeConverted)
```

Comment:
BTW, seems like you're mixing JDK 7 and 8 libraries which will likely be correct.

Reply:
Mixing is possible via the rebase methods like `RebaseDateTime.rebaseGregorianToJulianMicros()` otherwise an implementation works incorrectly for ancient dates (before 1582-10-15) and timestamps (before 1900-01-01) because of different underlying calendar systems (Java 8 - Proleptic Gregorian calendar, Java 7 - the hybrid calendar Julian + Gregorian calendar), and JDK implementations - Java 8 uses more precise historical data.

Since you mix Java 8 methods and Java 7 `Timestamp.valueOf()`, highly likely your implementation will work incorrectly.

Reply:
So what should I used to resolve this, any hints? Thanks

Reply:
Sorry, can you explain more, I thought Timestamp.valueOf() is from java 1.8?
![image](https://user-images.githubusercontent.com/84778052/119941256-8cae7d80-bfba-11eb-8a14-ea94b76df4dc.png)

Reply:
This method uses other methods of the `Timestamp` class that assumes the hybrid calendar. We described nuances of the hybrid and Proleptic Gregorian calendar (used since Spark 3.0) in the blog post https://databricks.com/blog/2020/07/22/a-comprehensive-look-at-dates-and-timestamps-in-apache-spark-3-0.html . I do believe that your changes can cause some issues in corner cases. I will try to allocate some time to find such cases slightly later.

Reply:
@MaxGekk any update?